plugin-api/src/main/java/org/sonar/api/batch/sensor/rule/AdHocRule.java

```diff
@@ -53,12 +56,29 @@ public interface AdHocRule {
 
   /**
    * Default severity of the rule.
+   * @deprecated since 10.1, use {@link #defaultImpacts()} instead
    */
+  @Deprecated(since = "10.1")
   Severity severity();
 
   /**
    * Type of the rule.
+   * @deprecated since 10.1, use {@link #defaultImpacts()} instead
    */
+  @Deprecated(since = "10.1")
   RuleType type();
```

Comment:
We deprecate type and severity in a few places, but we are not making it Nullable here. Should they be nullable now if we encourage people to use impacts?

In example cleanCodeAttribute has @CheckForNull annotation

Reply:
I agree with that, it makes sense since we want to populate those fields if they are not provided,  based on the impacts. (reverse mapping). Either impacts or old issue/type must be provided